crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,387 @@
+use reth_rpc_types::engine::PayloadStatusEnum;
+use reth_stages::Pipeline;
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc::UnboundedReceiver;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tracing::*;
+
+mod error;
+pub use error::BeaconEngineError;
+
+mod message;
+pub use message::BeaconEngineMessage;
+
+mod pipeline_state;
+pub use pipeline_state::PipelineState;
+
+#[derive(Debug, Default)]
+enum BeaconEngineAction {
+    #[default]
+    None,
+    RunPipeline,
+}
+
+impl BeaconEngineAction {
+    fn run_pipeline(&self) -> bool {
+        matches!(self, BeaconEngineAction::RunPipeline)
+    }
+}
+
+/// The beacon consensus engine is the driver that switches between historical and live sync.
+///
+/// TODO: add more docs
+#[must_use = "Future does nothing unless polled"]
+pub struct BeaconConsensusEngine<
+    DB: Database,
+    U: SyncStateUpdater,
+    C: Consensus,
+    EF: ExecutorFactory,
+> {
+    db: Arc<DB>,
+    pipeline_state: Option<PipelineState<DB, U>>,
+    blockchain_tree: BlockchainTree<DB, C, EF>,
+    message_rx: UnboundedReceiverStream<BeaconEngineMessage>,
+    forkchoice_state: Option<ForkchoiceState>,
+    next_action: BeaconEngineAction,
+}
+
+impl<DB, U, C, EF> BeaconConsensusEngine<DB, U, C, EF>
+where
+    DB: Database + Unpin + 'static,
+    U: SyncStateUpdater + Unpin + 'static,
+    C: Consensus,
+    EF: ExecutorFactory + 'static,
+{
+    /// Create new instance of the [BeaconConsensusEngine].
+    pub fn new(
+        db: Arc<DB>,
+        pipeline: Pipeline<DB, U>,
+        blockchain_tree: BlockchainTree<DB, C, EF>,
+        message_rx: UnboundedReceiver<BeaconEngineMessage>,
+    ) -> Self {
+        Self {
+            db,
+            pipeline_state: Some(PipelineState::Idle(pipeline)),
+            blockchain_tree,
+            message_rx: UnboundedReceiverStream::new(message_rx),
+            forkchoice_state: None,
+            next_action: BeaconEngineAction::RunPipeline,
+        }
+    }
+
+    /// Returns `true` if the pipeline is currently idle.
+    fn pipeline_is_idle(&self) -> bool {
+        self.pipeline_state.as_ref().expect("pipeline state is set").is_idle()
+    }
+
+    /// Set next action to [SyncControllerAction::RunPipeline] to indicate that
+    /// controller needs to run the pipeline as soon as it becomes available.
+    fn pipeline_run_needed(&mut self) {
+        self.next_action = BeaconEngineAction::RunPipeline;
+    }
+
+    /// Handle the forkchoice updated message.
+    fn on_forkchoice_updated(&mut self, state: ForkchoiceState) -> PayloadStatusEnum {
+        self.forkchoice_state = Some(state.clone());
+        if self.pipeline_is_idle() {
+            match self.blockchain_tree.make_canonical(&state.head_block_hash) {
+                Ok(_) => PayloadStatusEnum::Valid,
+                // TODO: handle/match error
+                Err(_error) => {
+                    self.pipeline_run_needed();
```

Comment:
Do we need to match on a specific error? 

Is it always going to be the case that `make_canonical` erroring  means we need to trigger the sync?